hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/search/ConfigSearch.java

```diff
@@ -0,0 +1,423 @@
+            @Override
+            public PNCounterConfig getDynamicConfig(ConfigurationService configurationService, String name) {
+                return configurationService.findPNCounterConfig(name);
+            }
+
+            @Override
+            public PNCounterConfig getStaticConfig(Config staticConfig, String name) {
+                return staticConfig.getPNCounterConfig(name);
+            }
+
+            @Override
+            public Map<String, PNCounterConfig> getStaticConfigs(Config staticConfig) {
+                return staticConfig.getPNCounterConfigs();
+            }
+        });
+    }
+
+    /**
+     * Factory method for creating {@link Searcher} instances
+     *
+     * @param cls class of the config we're looking for
+     * @param staticConfig hazelcast static config
+     * @param configurationService configuration service
+     * @param isStaticFirst <code>true</code> if {@link Searcher} should look into static config first, and
+     *                      <code>false</code> otherwise.
+     * @param <T> type of config class
+     * @return {@link Searcher} for the given config type and conditions
+     * @see com.hazelcast.spi.properties.GroupProperty#SEARCH_DYNAMIC_CONFIG_FIRST
+     */
+    public static <T extends IdentifiedDataSerializable> Searcher<T> with(Class<T> cls, Config staticConfig,
+                                                                        ConfigurationService configurationService,
+                                                                        boolean isStaticFirst) {
+        return isStaticFirst
+                ? new StaticFirstSearcher<T>(CONFIG_SUPPLIERS.get(cls), configurationService, staticConfig)
+                : new DynamicFirstSearcher<T>(CONFIG_SUPPLIERS.get(cls), configurationService, staticConfig);
+    }
+
+    /**
+     * Factory method for creating {@link Searcher} instances
+     *
+     * @param configSupplier config supplier
+     * @param staticConfig hazelcast static config
+     * @param configurationService configuration service
+     * @param isStaticFirst <code>true</code> if {@link Searcher} should look into static config first, and
+     *                      <code>false</code> otherwise.
+     * @param <T> type of config class
+     * @return {@link Searcher} for the given config type and conditions
+     * @see com.hazelcast.spi.properties.GroupProperty#SEARCH_DYNAMIC_CONFIG_FIRST
+     * @see ConfigSupplier
+     */
+    public static <T extends IdentifiedDataSerializable> Searcher<T> with(ConfigSupplier<T> configSupplier,
+                                                                          Config staticConfig,
+                                                                          ConfigurationService configurationService,
+                                                                          boolean isStaticFirst) {
+        return isStaticFirst
```

Comment:
I'm thinking, this will create new instances of searchers each time we need the configuration, which is when the structure is being initialised and possibly later on in it's lifecycle. OTOH the `isStaticFirst` value does not change during the lifecycle of the hazelcast instance. Can this be changed so we create the strategy when initialising the hazelcast instance?